File: euler/Python/Problem 1/1.py
"""
    This program is used to find the sum of the multiples of any numbers up to
    a certain limit. It can accept any number of multiples. Solution is found
    in polynomial time.

    :copyright: (c) 2013 by Moataz Elmasry
"""

def multiples_sum(multiples_list, limit):
    """ Finds sum of multiples up to limit.

    For example, for multiples list of [2,3] and a limit of 10, this function
    will sum the set (0,2,4,6,8,3,9), returning 32.

    Args:
        multiples_list: list of base multiples to be summed.
        limit: upper-bound for multiples to sum.

    Returns:
        Sum of the multiples.
    """

    # Initialize set that will contain all products
    products_set = set()

    # Populate products set
    for multiple in multiples_list:
        # Return new set, retaining all previous elements in products_set but
        # with new multiples added.
        products_set |= set(range(0, limit, multiple))

    # Sum products set to obtain final solution
    return sum(products_set)


# Prompt user for space-separated multiples
multiples_list = [int(x) for x in input("Enter multiples separated by "
                                        "space:\n").split()]

# Prompt user for threshold
limit = int(input("Now, enter the value of the limit:\n"))

print("\nThe answer is:\n", multiples_sum(multiples_list, limit))

File: euler/OCaml/Problem 1/1.ml
(* 
 *Finds the sum of two integer multiples below a certain limit. Program 
 *receives input as command line arguments where first argument is the first
 *multiple, second argument is the second multiple, and third argument is the
 *limit. Solution found in O(1) time complexity.
 *)

open Array;;
open Printf;;

(* Finds the sum of all real numbers up until the limit *)
let summation limit =
  limit *. ((limit +. 1.0) /. 2.0);;

(* Gives back the sum of all multiples of a number below a certain limit *)
let multiples_sum multiple limit =
  multiple *. summation(floor (limit /. multiple));;

let main () =

  (* Retrieve user input *)
  let [| multiple1; multiple2; limit |] = 
    map float_of_string (sub Sys.argv 1 3) in

  (* Find sum of each multiple and the sum of the product of both multiples *)
  let multiple1_sum = multiples_sum multiple1 (limit -. 1.0) in
  let multiple2_sum = multiples_sum multiple2 (limit -. 1.0) in
  let product_sum = multiples_sum (multiple1 *. multiple2) (limit -. 1.0) in
  let result = int_of_float (multiple1_sum +. multiple2_sum -. product_sum) in

  printf "The answer is: %d\n" result;
  exit 0;;

main ();;
